src/geometry/getPerspectiveWarp.ts

```diff
@@ -0,0 +1,166 @@
+import { Matrix, SingularValueDecomposition } from 'ml-matrix';
+
+import type { Point } from '../utils/geometry/points.js';
+
+/**
+ * Options for getPerspectiveWarp function.
+ */
+export interface GetPerspectiveWarpOptions {
+  /**
+   * The horizontal dimension (in pixels) of the final rectified rectangular image.
+   */
+  width?: number;
+  /**
+   * The vertical dimension (in pixels) of the final rectified rectangular image.
+   */
+  height?: number;
+}
+/**
+ * Returns result matrix along with vertical and horizontal dimensions for the rectangular image.
+ */
+export type GetPerspectiveWarpData = Required<GetPerspectiveWarpOptions> & {
+  matrix: number[][];
+};
+
+// REFERENCES :
+// https://stackoverflow.com/questions/38285229/calculating-aspect-ratio-of-perspective-transform-destination-image/38402378#38402378
+// http://www.corrmap.com/features/homography_transformation.php
+// https://ags.cs.uni-kl.de/fileadmin/inf_ags/3dcv-ws11-12/3DCV_WS11-12_lec04.pdf
+// http://graphics.cs.cmu.edu/courses/15-463/2011_fall/Lectures/morphing.pdf
+/**
+ * Returns perspective warp matrix from 4 points.
+ * @param pts - 4 reference corners of the new image.
+ * @param options - PerspectiveWarpOptions
+ * @returns - Matrix from 4 points.
+ */
+export function getPerspectiveWarp(
+  pts: Point[],
+  options: GetPerspectiveWarpOptions = {},
+): GetPerspectiveWarpData {
+  if (pts.length !== 4) {
+    throw new Error(
+      `The array pts must have four elements, which are the four corners. Currently, pts have ${pts.length} elements`,
+    );
+  }
+  const { width, height } = options;
+  const [tl, tr, br, bl] = order4Points(pts);
+
+  let widthRect;
+  let heightRect;
+  if (height && width) {
+    widthRect = width;
+    heightRect = height;
+  } else if (!height && !width) {
+    widthRect = Math.ceil(
+      Math.max(distance2Points(tl, tr), distance2Points(bl, br)),
+    );
+    heightRect = Math.ceil(
+      Math.max(distance2Points(tl, bl), distance2Points(tr, br)),
+    );
+  } else {
+    throw new Error(`${width ? 'Height' : 'Width'} is not defined.`);
```

Comment:
You can make a better error message:
- Be more explicit and complete. Explain the options' `width` and `height` properties are invalid with an explanation on how to use them correctly together.
- When using property names like `width` and `height` in an error message, I would advise to backquote them and respect the casing.